stdlib/LinearAlgebra/src/bidiag.jl

```diff
@@ -409,7 +409,10 @@ function diag(M::Bidiagonal{T}, n::Integer=0) where T
     elseif (n == 1 && M.uplo == 'U') ||  (n == -1 && M.uplo == 'L')
         return copyto!(similar(M.ev, length(M.ev)), M.ev)
     elseif -size(M,1) <= n <= size(M,1)
-        v = similar(M.dv, size(M,1)-abs(n))
+        # We compute the element type as zero(M[1,1]) if !haszero(T)
+        # This only works as we limit n to abs(n) <= size(M,1), instead of allowing arbitrary values
+        # The limit ensures that n>0 is allowed only if M has at least one element
+        v = similar(M.dv, typeof(_zero(M, 1, 1)), size(M,1)-abs(n))
         for i in eachindex(v)
             v[i] = M[BandIndex(n,i)]
         end
```

Comment:
Even better: use a comprehension?

```suggestion
        v = Base.collect_similar(M, (M[BandIndex(n,i)] for i in eachindex(v)))
```

Reply:
I like this approach, but it seems to perform slightly worse at type-inference for the `static` case, as the element type isn't inferred if there are no elements.
```julia
julia> B = Bidiagonal(fill(static(1),4), fill(static(1),3), :U)
4×4 Bidiagonal{StaticInt{1}, Vector{StaticInt{1}}}:
 static(1)  static(1)      ⋅          ⋅    
     ⋅      static(1)  static(1)      ⋅    
     ⋅          ⋅      static(1)  static(1)
     ⋅          ⋅          ⋅      static(1)


julia> @code_warntype (B -> diag(B,2))(B)
MethodInstance for (::var"#1#2")(::Bidiagonal{StaticInt{1}, Vector{StaticInt{1}}})
  from (::var"#1#2")(B) @ Main REPL[9]:1
Arguments
  #self#::Core.Const(var"#1#2"())
  B::Bidiagonal{StaticInt{1}, Vector{StaticInt{1}}}
Body::Union{Vector{StaticInt{1}}, Vector{StaticInt}, Vector{StaticInt{0}}}
1 ─ %1 = Main.diag(B, 2)::Union{Vector{StaticInt{1}}, Vector{StaticInt}, Vector{StaticInt{0}}}
└──      return %1
```
It's a bit of a shame that constant-propagation doesn't stabilize the types here.

Reply:
That use-case seems likely to be broken in other ways, though — it reports the eltype as `StaticInt{1}` when it should be `Union{StaticInt{0}, StaticInt{1}}`.

Trying to guess based upon the value of the first element is likely to go awry.  I suspect doing this breaks in ways that would result in correctness problems.

```julia
julia> D = Diagonal([missing, 2, 3, 4])
4×4 Diagonal{Union{Missing, Int64}, Vector{Union{Missing, Int64}}}:
  missing  ⋅  ⋅  ⋅
 ⋅         2  ⋅  ⋅
 ⋅         ⋅  3  ⋅
 ⋅         ⋅  ⋅  4

julia> diag(D, 2)
2-element Vector{Union{Missing, Int64}}:
 0
 0
```